src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.MessageMode.Windows.cs

```diff
@@ -0,0 +1,106 @@
+
+            Assert.Equal(5, bytesRead);
+            Assert.Equal(new byte[] { 1, 2, 3, 4, 5, 0 }, buffer1.Concat(buffer2).Concat(buffer3));
+            Assert.Equal(new bool[] { false, false, true }, messageCompleted);
+
+            await serverWrite;
+        }
+
+        [Theory]
+        [InlineData(PipeTransmissionMode.Byte, PipeOptions.None)]
+        [InlineData(PipeTransmissionMode.Message, PipeOptions.None)]
+        [InlineData(PipeTransmissionMode.Byte, PipeOptions.Asynchronous)]
+        [InlineData(PipeTransmissionMode.Message, PipeOptions.Asynchronous)]
+        public void ServerIn_ClientConnect_Throws(PipeTransmissionMode serverMode, PipeOptions options)
+        {
+            string pipeName = PipeStreamConformanceTests.GetUniquePipeName();
+
+            using NamedPipeServerStream server = new NamedPipeServerStream(pipeName, PipeDirection.In, 1, serverMode, options);
+            using NamedPipeClientStream client = CreateClientStream(pipeName, options);
+
+            Assert.Throws<UnauthorizedAccessException>(() => client.Connect());
+        }
+
+        [Theory]
+        [InlineData(PipeDirection.Out, PipeTransmissionMode.Byte, PipeOptions.None)]
+        [InlineData(PipeDirection.Out, PipeTransmissionMode.Byte, PipeOptions.Asynchronous)]
+        [InlineData(PipeDirection.InOut, PipeTransmissionMode.Byte, PipeOptions.None)]
+        [InlineData(PipeDirection.InOut, PipeTransmissionMode.Byte, PipeOptions.Asynchronous)]
+        public async Task ServerByteMode_ClientReadModeMessage_Throws(PipeDirection serverDirection, PipeTransmissionMode serverMode, PipeOptions options)
+        {
+            string pipeName = PipeStreamConformanceTests.GetUniquePipeName();
+
+            using NamedPipeServerStream server = new NamedPipeServerStream(pipeName, serverDirection, 1, serverMode, options);
+            using NamedPipeClientStream client = CreateClientStream(pipeName, options);
+
+            Task serverConnected = server.WaitForConnectionAsync();
+            client.Connect();
+            await serverConnected;
+
+            Assert.Throws<IOException>(() => client.ReadMode = PipeTransmissionMode.Message);
+        }
+
+        [Fact]
+        public async Task PipeAccessRights_Without_WriteAttributes_ClientReadModeMessage_Throws()
+        {
+            string pipeName = PipeStreamConformanceTests.GetUniquePipeName();
+            PipeAccessRights rights = MinimumMessageAccessRights & ~PipeAccessRights.WriteAttributes;
+
+            using NamedPipeServerStream server = new NamedPipeServerStream(pipeName, PipeDirection.InOut, 1, PipeTransmissionMode.Message);
+            using NamedPipeClientStream client = new NamedPipeClientStream(".", pipeName, rights, PipeOptions.None, Security.Principal.TokenImpersonationLevel.None, HandleInheritability.None);
+
+            Task serverConnected = server.WaitForConnectionAsync();
+            client.Connect();
+            await serverConnected;
```

Comment:
Ditto.

It would be nice to see synchronous methods and asynchronous methods doing the same logic. Similar to System.Formats.Tar: we have files to test the async methods only, and files to test the sync methods only, but never intermixed. Here we don't need separate files, they can be added to the same file. The async methods need to end with the *Async suffix.

Reply:
These tests are focused on the message mode itself, not the aspect of making connections, so I'll leave that out of scope to ensure all combinations of sync/async connections are covered. But I'll standardize onto `Task.WaitAll(server.WaitForConnectionAsync(), client.ConnectAsync())` instead of mixing sync/async to make it more readable.

Since the access rights overload throws on non-Windows platforms, I'll keep these tests in a separate file too (otherwise, they'd need to all be `[PlatformSpecific(TestPlatforms.Windows)]`).